### PR TITLE
support a coroutine function in t.i.task.react

### PR DESCRIPTION
**Pull request by graingert**
## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/9974
* [x] I ran `tox -e black-reformat` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I ran additional checks listed at [Getting Your Patch Accepted](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

### CODE REVIEW

File: src/twisted/internet/task.py [hunk 1]
```diff
@@ -860,14 +861,14 @@ def react(main, argv=(), _reactor=None):
 
     The following demonstrates the signature of a C{main} function which can be
     used with L{react}::
-          def main(reactor, username, password):
-              return defer.succeed('ok')
+          async def main(reactor, username, password):
+              return "ok"
```
Comment (wsanchez) [inline src/twisted/internet/task.py:865]: yes, please!

File: src/twisted/internet/task.py [hunk 2]
```diff
@@ -860,14 +861,14 @@ def react(main, argv=(), _reactor=None):
...
 
-          task.react(main, ('alice', 'secret'))
+          task.react(main, ("alice", "secret"))
 
-    @param main: A callable which returns a L{Deferred}. It should
-        take the reactor as its first parameter, followed by the elements of
-        C{argv}.
+    @param main: A callable which returns a L{Deferred} or
+        C{async def} C{coroutine}. It should take the reactor as its first
```
Comment (wsanchez) [inline src/twisted/internet/task.py]: Instead of "C{async def} C{coroutine}" I'd just say "coroutine".

File: src/twisted/test/test_task.py [hunk 3]
```diff
@@ -1205,3 +1205,214 @@ def main(reactor):
         r = _FakeReactor()
         exitError = self.assertRaises(SystemExit, task.react, main, [], _reactor=r)
         self.assertEqual(0, exitError.code)
+
+
+class ReactCoroFnTests(unittest.SynchronousTestCase):
```
Comment (wsanchez) [inline src/twisted/test/test_task.py]: Readability matters, so maybe rename `ReactCoroFnTests` to `ReactCoroutineFunctionTests`. Length here is harmless.

**Review by wsanchez — APPROVED** [review]
Awesome.

Minor suggestions here, but good to go.

File: src/twisted/test/test_task.py (hunk 3 above)
Comment (graingert) [reply src/twisted/test/test_task.py]:
```suggestion
class ReactCoroutineFunctionTests(unittest.SynchronousTestCase):
```

File: src/twisted/internet/task.py (hunk 2 above)
Comment (graingert) [reply src/twisted/internet/task.py]:
```suggestion
        coroutine. It should take the reactor as its first
```

**Comment by wsanchez** [comment]
@graingert looks good to me

**Pull request merged**